### DISCUSSION TITLE
Need a little bit of help with OOBE\BYPASSNRO

### DISCUSSION THREAD

**Discussion by ghost** [AveYo/MediaCreationTool.bat#275] [General]
Hello all, I like playing with ISO files and I'm trying to create my ideal Windows 11 ISO on my own.

As of now I figured out how to make the ISO bypass TPM/Secure Boot/RAM, etc., and activate Windows automatically.

What I'm trying to do (already read the source code of the project in the section that bypasses Microsoft Account requirement) is set the **OOBE\BYPASSNRO** in the ISO so I don't have to type it via CMD and restart the OOBE. Read a little bit the code MCT uses and saw the registry value it adds, but adding it on my own the same way I added the TPM bypass didn't work.

Is there any tutorial or can someone provide me with some information on how to only add the **OOBE\BYPASSNRO** in the ISO so I don't have to type it via CMD when installing Windows? I also found [this](https://github.com/pbatard/rufus/issues/1981#issue-1301353017) but I don't know how to do it.

TIA, cheers!

**Comment by AveYo** [comment]
All you need is already provided in [bypass11/AutoUnattend.xml](https://github.com/AveYo/MediaCreationTool.bat/blob/main/bypass11/AutoUnattend.xml)  
Just include it in your iso sources folder and you're all set!  
Tho it's most convenient if it's added inside `boot.wim` like it's done in MediaCreationTool.bat so that it does not interfere with upgrades under windows.

I don't think you care much about the never used, rarely if ever working Repair my PC under winpe.  
And I'm sure I can address that too if I put my mind to it.
When it comes to bypassing 11 requirements, this project offered both the bleeding edge and the most convenient workarounds and others have taken inspiration from it, not the other way around ;) 

Haven't done a fresh test in a long time, but once I do I will give you a minimalist example

**Comment by ghost** [comment]
> All you need is already provided in [bypass11/AutoUnattend.xml](https://github.com/AveYo/MediaCreationTool.bat/blob/main/bypass11/AutoUnattend.xml)  
> Just include it in your iso sources folder and you're all set!  
> Tho it's most convenient if it's added inside `boot.wim` like it's done in MediaCreationTool.bat so that it does not interfere with upgrades under windows.
> 
> I don't think you care much about the never used, rarely if ever working Repair my PC under winpe.  
> And I'm sure I can address that too if I put my mind to it.
> When it comes to bypassing 11 requirements, this project offered both the bleeding edge and the most convenient workarounds and others have taken inspiration from it, not the other way around ;) 
> 
> Haven't done a fresh test in a long time, but once I do I will give you a minimalist example

Thank you for the answer @AveYo! I thought I had to put the AutoUnattend.xml file in the root of the ISO folder and not in the Sources folder, since all tutorials I saw put it there.

I also added my own $OEM$ folder to activate Windows 11 automatically under the Sources folder and changed the Product Key in your AutoUnattend.xml file to match the one in the $OEM$.

I just started booting the ISO and will se what happens. 2 questions:

1. Do I need to have the ethernet unplugged for the BYPASSNRO to work correctly?
2. Can you elaborate on what you mean by "_so that it does not interfere with upgrades under windows._"?

Am I going to have problems with Windows updates and such?

**Comment by ghost** [comment]
Update: @AveYo I reached the Microsoft Account requirement and I got an error that said "Oops something went wrong OOBEBYPASSNRO".

Is this because I had to put AutoUnattend.xml in the Sources folder, or it's because I didn’t unplug ethernet?

**Comment by ghost** [comment]
UPDATE2: Without the ethernet plugged in, it worked correctly. I guess that was the problem or I was lucky and the second time without ethernet it happened to show the Skip button? 

Also could you please clarify what you ment about problems with upgrades in Windows?

TIA, @AveYo !

**Comment by AveYo** [comment]
AutoUnattend.xml placed on root of media or on sources folder will work fine when booting from media  
but under windows, when you run setup.exe for upgrade / repair it will launch the same setup as boot (sources/setup.exe) and that one can only install on another hdd / partition, not do upgrades.  

In MediaCreationTool.bat it's placed instead in boot.wim, index 2 at the root of mounted image, so it will work fine when booting from media but also under windows, where setup will launch the proper sources/setupprep.exe (which you can do manually with the former if you know about it).

You can always try it out yourself, don't take my word for it.

**Comment by ghost** [comment]
> AutoUnattend.xml placed on root of media or on sources folder will work fine when booting from media  
> but under windows, when you run setup.exe for upgrade / repair it will launch the same setup as boot (sources/setup.exe) and that one can only install on another hdd / partition, not do upgrades.  
> 
> In MediaCreationTool.bat it's placed instead in boot.wim, index 2 at the root of mounted image, so it will work fine when booting from media but also under windows, where setup will launch the proper sources/setupprep.exe (which you can do manually with the former if you know about it).
> 
> You can always try it out yourself, don't take my word for it.

Now it's more clear.

So this means that for a clean install it's completely fine and I got the error that didn't allow me to skip Microsoft Account because I indeed had ethernet plugged in? There is no way to have the Skip button show while ethernet is plugged in (can't use this to bypass Microsoft Account when installing in a laptop?)?

BTW sorry for the spam haha, I love this project! I don't even use Windows for a daily driver and I just learn all this because it very interesting. I provide these ISOs to family and friends!

**Reply by AveYo** [reply]
The issue was setup not allowing you to continue without an internet connection in Home (& Pro) editions.
If you're already connected at setup (like via ethernet cable) the **I don't have internet** won't show up.  
In that case it's well known you can use user `a` password `a` (and many such) to skip MSA to a local account.

I just tested a 11 Home, Canary build 25905.1000 with the only modification of adding AutoUnattend.xml to the boot.wim index 2. Requirements were bypassed, `I don't have internet` appeared (can just disconnect at that screen and restart setup, it will appear).

I like workarounds that stand the test of time (last modified March last year).

**Reply by AveYo** [reply]
Also, the requirements bypass is harmless on fully supported machines, so better to have it than not for the chance that something happens and TPM is disabled in BIOS or something.

**Reply by ghost** [reply]
> The issue was setup not allowing you to continue without an internet connection in Home (& Pro) editions.
> If you're already connected at setup (like via ethernet cable) the **I don't have internet** won't show up.  
> In that case it's well known you can use user `a` password `a` (and many such) to skip MSA to a local account.
> 
> I just tested a 11 Home, Canary build 25905.1000 with the only modification of adding AutoUnattend.xml to the boot.wim index 2. Requirements were bypassed, `I don't have internet` appeared (can just disconnect at that screen and restart setup, it will appear).
> 
> I like workarounds that stand the test of time (last modified March last year).
> 

Excellent.

Is there any easier way to add AutoUnattend.xml to the boot.wim without having to mount and run all those commands in CMD? How did you just add AutoUnattend.xml in boot.wim on a fresh Win11 ISO?

I know MCT does this, but since I'm trying to learn I want to do it manually, even though I still use MCT's AutoUnattend.xml because you're god damn right that all other projects use MCT's scripts and not the other way around ;)

**Reply by AveYo** [reply]
It takes the longest but it's the safest method to use dism and mount the boot.wim - and it can be batch scripted easily:

```bat
dism.exe /mount-wim /wimfile:"AN_ISO_TO_FOLDER\sources\boot.wim" /index:2 /mountdir:"AN_EMPTY_TEMP_DIR"
copy /y "PATH_TO_PROJECT\AutoUnattend.xml" "AN_EMPTY_TEMP_DIR\"
rem pause - could wait here and do more file operations manually, then enter to resume
dism.exe /unmount-wim /mountdir:"AN_EMPTY_TEMP_DIR" /commit
rmdir /s /q "AN_EMPTY_TEMP_DIR"
```

3rd-party dism alternative like [wimlib](https://wimlib.net) can directly add files like in an archive, offers the needed documentation, it's non-gui so also scriptable. usually faster if you skip re-optimization.  

I casually use total commander, so extracting iso to a folder and copy files from one source panel to another where I've opened an iso or a wim like an archive via 7-zip plugin is like second nature.
Then take advantage of button bars / sendto context menu scripts / context menu extensions like my MakeISO to turn a folder back into an iso ( _originally shared on mydigitallife - I've lost access to my fav community :(_ ) but here's the whole tiny code, save as .reg and import:

```reg
Windows Registry Editor Version 5.00

; Directory context menu for MakeISO - lean and mean snippet by AveYo, 2022.05.04

[-HKEY_CLASSES_ROOT\MakeISO]
[-HKEY_CLASSES_ROOT\Directory\shell\extract]
; To remove entries, copy paste above into undo_MakeISO.reg file, then import it

; MakeISO on Folder
[HKEY_CLASSES_ROOT\Directory\shell\extract]
"MuiVerb"="MakeISO"
"Icon"="imageres.dll,-5205"
"MultiSelectModel"="Single"
"AppliesTo"="NOT (System.Kind:~Library) AND System.Kind:~Folder"

[HKEY_CLASSES_ROOT\Directory\shell\extract\command]
@="C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe -nop -c iex((10..31|%%{(gp 'Registry::HKCR\\MakeISO' $_ -ea 0).$_})-join[char]10); # --%% \"%L\""

; MakeISO function
[HKEY_CLASSES_ROOT\MakeISO]
"10"="function MakeISO ($dir,$iso,$label='DVD_ROM') {if (!(test-path -Path $dir -pathtype Container)) {\"[ERR] $dir\"; return 1}; $code=@\""
"11"=" using System; using System.IO; using System.Runtime.Interop`Services; using System.Runtime.Interop`Services.ComTypes;"
"12"=" public class dir2iso {public int AveYo=2021; [Dll`Import(\"shlwapi\",CharSet=CharSet.Unicode,PreserveSig=false)]"
"13"=" internal static extern void SHCreateStreamOnFileEx(string f,uint m,uint d,bool b,IStream r,out IStream s);"
"14"=" public static int Create(string file, ref object obj, int bs, int tb) { IStream dir=(IStream)obj, iso;"
"15"=" try {SHCreateStreamOnFileEx(file,0x1001,0x80,true,null,out iso);} catch(Exception e) {Console.WriteLine(e.Message); return 1;}"
"16"=" int d=tb>1024 ? 1024 : 1, pad=tb%d, block=bs*d, total=(tb-pad)/d, c=total>100 ? total/100 : total, i=1, MB=(bs/1024)*tb/1024;"
"17"=" Console.Write(\"...  {0}MB \",MB); if (pad > 0) dir.CopyTo(iso, pad * block, Int`Ptr.Zero, Int`Ptr.Zero);"
"18"=" while (total-- > 0) {dir.CopyTo(iso, block, Int`Ptr.Zero, Int`Ptr.Zero); if (total % c == 0) {Console.Write(\"\\r{0,2}%\",i++);}}"
"19"=" iso.Commit(0); Console.WriteLine(\"\\r{0,2}%  {1}MB \",100,MB); return 0;} } // ` used to silence ps eventlog"
"20"="\"@; \"MakeISO $dir\"; & { $cs = new-object CodeDom.Compiler.CompilerParameters; $cs.GenerateInMemory = 1"
"21"=" $compile = (new-object Microsoft.CSharp.CSharpCodeProvider).CompileAssemblyFromSource($cs, $code)"
"22"=" $BOOT = @(); $bootable = 0; $mbr_efi = @(0,0xEF); $images = @('boot\\etfsboot.com','efi\\microsoft\\boot\\efisys.bin') #:: _noprompt"
"23"=" 0,1|% { $bootimage = join-path $dir -child $images[$_]; if (test-path -Path $bootimage -pathtype Leaf) {"
"24"=" $bin = new-object -ComObject ADODB.Stream; $bin.Open(); $bin.Type = 1; $bin.LoadFromFile($bootimage)"
"25"=" $opt = new-object -ComObject IMAPI2FS.BootOptions;$opt.AssignBootImage($bin.psobject.BaseObject); $opt.PlatformId = $mbr_efi[$_]"
"26"=" $opt.Emulation = 0; $bootable = 1; $opt.Manufacturer = 'Microsoft'; $BOOT += $opt.psobject.BaseObject } }"
"27"=" $fsi = new-object -ComObject IMAPI2FS.MsftFileSystemImage; $fsi.FileSystemsToCreate = 4; $fsi.FreeMediaBlocks = 0"
"28"=" if ($bootable) {$fsi.BootImageOptionsArray = $BOOT}; $TREE = $fsi.Root; $TREE.AddTree($dir,$false); $fsi.VolumeName = $label"
"29"=" $obj = $fsi.CreateResultImage(); $ret = [dir2iso]::Create($iso,[ref]$obj.ImageStream,$obj.BlockSize,$obj.TotalBlocks) }"
"30"=" [GC]::Collect(); return $ret # export directory as (bootable) udf iso - lean and mean snippet by AveYo, 2022.05.04"
"31"="}; $A=([environment]::commandline-split'-[-]%+ ?',2)[1].Trim(' \"'); $L=split-path -leaf $A; MakeISO $A \"$L.iso\" $L"
```

Total commander is not a must, standalone 7-zip is usable too directly from explorer to modify wim's.
Ultra-Iso is also good, can accelerate iso operations (don't even need to recreate the whole iso, just extract boot.wim, patch it and replace it in-place).  

Under linux it's mkisofs + wimlib so kinda similar.

I salute you for your tinkering passion, not too long until it's gonna become a lost art and most gonna be at the mercy of greedy corporations colluding among themselves against the user's best interests. 
Don't have much hope from the "can you make a video of x?" crowd - well duh! you wanna see me copy-paste the exact few lines in powershell?! But I've seen promising minecraft kids ;)

**Reply by ghost** [reply]
@AveYo thank you so much for all the info you provided and the salute haha!

Open projects like this are the inspiration for me to tinker and learn more. When you have projects that all they do is provide solutions "for the people" and not corporations that a lot of times do worse job than a collective of people with passion for open-source, it's difficult not to grt inspired with and try yourself ;)
